### PR TITLE
DigiSubs2023 price migration

### PR DESCRIPTION
**Pull request by shtukas**
We implement the engine update for the DigiSubs2023 migration.

**Part 1**

This PR has the following preliminaries (refactoring PRs)
- https://github.com/guardian/price-migration-engine/pull/936
- https://github.com/guardian/price-migration-engine/pull/937
- https://github.com/guardian/price-migration-engine/pull/941 

And one config PR
- https://github.com/guardian/membership-workflow/pull/499

With the refactorings we are now isolating in specific files (a convention we started with Membership, and which has been applied to GuardianWeekly and SupporterPlus) but also a specific namespace. We are also introducing a new space for migration specific test files. This enable us to neatly separate generic models on one side and migration specific code on the other, allowing us to better decommission migration code in the future.

**Part 2**

The DigiSubs2023 migration is structurally simple: we remain within similar rate plans and simply update the prices. Note that we do not apply price capping there.

### CODE REVIEW

**Comment by shtukas** [comment]
Double checking the rate plan ids mapping

Screenshot 2023-10-09 at 11 14 12

Screenshot 2023-10-09 at 11 14 25

Screenshot 2023-10-09 at 11 14 31

```
private val billingPeriodToRatePlanId: Map[BillingPeriod, TargetRatePlanDetails] = Map(
  BillingPeriod.Month -> TargetRatePlanDetails(
    "2c92a0fb4edd70c8014edeaa4eae220a",
    "2c92a0fb4edd70c9014edeaa50342192"
  ), // Digital Pack Monthly (rate plan) (old)
  BillingPeriod.Quarterly -> TargetRatePlanDetails(
    "2c92a0fb4edd70c8014edeaa4e8521fe",
    "2c92a0fb4edd70c9014edeaa4fd42186"
  ), // Digital Pack Quarterly (rate plan) (old)
  BillingPeriod.Annual -> TargetRatePlanDetails(
    "2c92a0fb4edd70c8014edeaa4e972204",
    "2c92a0fb4edd70c9014edeaa5001218c"
  ), // Digital Pack Annual (rate plan) (old)
)
```

File: lambda/src/main/scala/pricemigrationengine/handlers/EstimationHandler.scala [hunk 1]
```diff
@@ -1,5 +1,6 @@
 package pricemigrationengine.handlers
 
+import pricemigrationengine.migrations.DigiSubs2023Migration
```
Comment (kelvin-chappell) [inline lambda/src/main/scala/pricemigrationengine/handlers/EstimationHandler.scala:3]: Not being used

File: lambda/src/main/scala/pricemigrationengine/migrations/DigiSubs2023Migration.scala [hunk 2]
```diff
@@ -0,0 +1,251 @@
...
+          case Some(price) => Right(price)
+        }
+    }
+  }
+
+  def subscriptionRatePlan(subscription: ZuoraSubscription): Either[AmendmentDataFailure, ZuoraRatePlan] = {
+    // Takes a subscription and return the active rate plan
+    // This function is specific to the current Migration, eg: DigiSubs2023, so we can be effective in the look up of
+    // that rate plan
+    subscription.ratePlans.filter(rp => rp.productName == "Digital Pack").headOption match {
```
Comment (kelvin-chappell) [inline lambda/src/main/scala/pricemigrationengine/migrations/DigiSubs2023Migration.scala:105]:
```suggestion
    subscription.ratePlans.find(_.productName == "Digital Pack") match {
```

File: lambda/src/test/resources/DigiSubs2023/readme.txt [hunk 3]
```diff
@@ -0,0 +1 @@
+We use `regular` for the standard subscription we are going to use to test all the migration basic functions and `discounted` for the discounted subscription that we are also using to test the detection of discounted subscriptions that need to be ignored.
```
Comment (kelvin-chappell) [inline lambda/src/test/resources/DigiSubs2023/readme.txt:1]: Might be better for maintainability to put this into a readme.md at the top of the test resources as a general explanation of structure.  And then link to it from the project readme.

File: lambda/src/main/scala/pricemigrationengine/model/BillingPeriod.scala [hunk 4]
```diff
@@ -0,0 +1,26 @@
+package pricemigrationengine.model
+
+object BillingPeriod extends Enumeration {
+  type BillingPeriod = Value
+  val Month, Quarterly, Annual = Value
```
Comment (kelvin-chappell) [inline lambda/src/main/scala/pricemigrationengine/model/BillingPeriod.scala:5]: I thought there were model billing periods than this, like semiannual, unless they have all been normalised, which would make things easier

Comment (shtukas) [reply lambda/src/main/scala/pricemigrationengine/model/BillingPeriod.scala:5]: Fixed here: https://github.com/guardian/price-migration-engine/pull/941

File: lambda/src/test/resources/DigiSubs2023/readme.txt (hunk 3 above)
Comment (shtukas) [reply lambda/src/test/resources/DigiSubs2023/readme.txt:1]: Actually that explanation is only for DigiSubs2023. For that particular subscription I needed to have a discounted subscription (in addition to the standard one)

File: lambda/src/main/scala/pricemigrationengine/migrations/DigiSubs2023Migration.scala (hunk 2 above)
Comment (shtukas) [reply lambda/src/main/scala/pricemigrationengine/migrations/DigiSubs2023Migration.scala]: https://github.com/guardian/price-migration-engine/pull/938/commits/2dd14d18dd35b6d04cb7f9cbf6395bdd459667df
Merci

File: lambda/src/main/scala/pricemigrationengine/handlers/EstimationHandler.scala (hunk 1 above)
Comment (shtukas) [reply lambda/src/main/scala/pricemigrationengine/handlers/EstimationHandler.scala:3]: Actually it was being used, here: https://github.com/guardian/price-migration-engine/pull/938/commits/9ff40c65ccd4ce5de6569c0176c335cffd1517ed#diff-23bf94a527ab471ddb3378219c90b6bc237922cbc47e53ea002e247b611a5e6eR84

But I have just removed that line, so it's not being use anymore :)

**Review by kelvin-chappell — APPROVED** [review]
Looks good! 🚀

**Pull request merged**